### ISSUE TITLE
Wrong colors in composite with blend:'dest-in' + resize since v0.30.2

### ISSUE THREAD

**Issue by tophf** [lovell/sharp#3334]
- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.
- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```
  System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz
    Memory: 8.27 GB / 31.95 GB
  Binaries:
    Node: 16.7.0 - C:\app\dev\nodejs\node.EXE
    Yarn: 1.22.17 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 7.20.3 - C:\app\dev\nodejs\npm.CMD
  npmPackages:
    sharp: ^0.30.7 => 0.30.7
```

### What are the steps to reproduce?

1. Resize a semi-transparent image using the code below
2. The result is an overexposed image

### What is the expected behaviour?

The image is half-transparent as shown in the table below.

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem

```js
require('sharp')('icon.png').composite([{
  input: Buffer.from([255, 255, 255, 256 * 0.5]),
  raw: { width: 1, height: 1, channels: 4 },
  tile: true,
  blend: 'dest-in',
}]).resize({ width: 64 }).toFile('icon-t.png');
```
Type | Source | Expected | Observed
-|-|-|-
Image | ![icon](https://user-images.githubusercontent.com/1310400/184409193-c801f22f-e427-4d6d-b8b0-973a70fc7011.png) | ![icon-t](https://user-images.githubusercontent.com/1310400/184409506-5e4b93ba-78cb-49db-9e37-494c3d80e522.png) | ![icon-t](https://user-images.githubusercontent.com/1310400/184409202-67266c87-3f1b-4ae7-8699-665501356a6f.png)

Looking at https://github.com/lovell/sharp/compare/v0.30.1...v0.30.2 I guess the bug was introduced in c62002554b5938c43c14fefafad7353ae4d720cf released in 0.30.2. Further investigation showed it was partly fixed - for non-resized composite - in 23033e2050d9ef099ecb02dde0f56f9f23e8d30d released in 0.30.3.

**Comment by lovell** [comment]
Thanks for reporting this, the changes to composite you discovered did cause this. Commit https://github.com/lovell/sharp/commit/a618ce7a1593e3dcc7d2d2170477389efd5fe420 now ensures the (implicit) unpremultiply operation occurs before composite and adds a test that would previously have failed. This will be included in v0.31.0.

**Comment by lovell** [comment]
v0.31.0 now available with the fix, thanks for reporting this.
